### PR TITLE
Update bitcoin.service to conform to init.md

### PR DESCRIPTION
**Pull request by dongcarl**
- `-datadir` option specified.
- Ask systemd to create and set the right mode for PID directory, configuration directory, and data directory.
- Tell systemd our group so it will set the right owner for aforementioned directories.

More information: https://www.freedesktop.org/software/systemd/man/systemd.exec.html

### CODE REVIEW

**Comment by laanwj** [comment]
Last update was in PR #10529 - please try to coordinate the changes with the people replying there.

**Comment by dongcarl** [comment]
@laanwj Sorry I'm not sure I entirely know what you mean. Am I to bring up this PR on that thread or shall I tag them in a comment here?

**Comment by MarcoFalke** [comment]
I think the link was supposed to be  #12102

**Comment by laanwj** [comment]
> Sorry I'm not sure I entirely know what you mean. Am I to bring up this PR on that thread or shall I tag them in a comment here?

Either, or both, is fine!

**Comment by dongcarl** [comment]
@laanwj @MarcoFalke should I be waiting for the other PR to get merged? Seems like the OP has been unresponsive.

**Comment by MarcoFalke** [comment]
I think the only thing left is give people time to review and then merge both whenever they are reviewed.

**Comment by dongcarl** [comment]
Cool beans.

**Comment by Flowdalic** [comment]
Thanks for your ping in https://github.com/bitcoin/bitcoin/pull/12102#issuecomment-362968939, here are my thoughts on this PR:

This PR contains the following changes to the systemd bitcoind.service file
1. Provide `-datadir` argument to bitcoind via ExecStart
2. Create various runtime directories and set their access mode
3. Set the Group to bitcoin

I deliberately did not put (1) in the systemd file as it can also be set via bitcoind's conf. I believe that it is good practice to put as much as possible in generic service config file (here: bitcoin.conf) as opposed to the system's service management facilities (here: systemd).

(2) is typically done by the distribution's package manager and would hence override settings of distributions that don't use those standard directories (therefore those distributions would need to patch the upstream systemd file or use their own, which is not both not desirable). Nit: I don't think it is necessary to protect the runtime directory with 700 as there a no secrets in it.

(3) is changes the behavior as it explicitly mentions the existence of a group named 'bitcoin', which may or may not exist. I believe that the behavior if 'Group' is not set, using the default group of the user, is more portable and should be used.

Further nits of this PR include that the order of `-pid` and `-conf` is changed for no reason, and the (correct)  comment telling us that the `/run/bitcoind` is owned by 'bitcoin' is removed.

In conclusion I am not provided with and don't see any compelling reasons to apply this.

**Comment by randolf** [comment]
utACK.

**Comment by sugarjig** [comment]
I'd like to throw in a vote for (1). I recently stood up an Ubuntu 18.04 server, using the instructions in `doc/init.md` to install the service. Like @dongcarl is saying, I was a bit confused when it came to configuring the data directory. It worked when I added the `-datadir` option to the service file.

I tried adding `datadir=/var/lib/bitcoind` to `bitcoin.conf` like @Flowdalic suggested, but Bitcoin failed to start. Running `journalctl -xe`, I see the the error `boost::filesystem::create_directories: Permission denied: "/home/bitcoin"`. It seems this option in the config file has no effect.

**Comment by Flowdalic** [comment]
I had a look at the code and it appears that bitcoind init first checks the datadir and then reads the config file, which would explain why `datadir` in the config file has no effect if the default datadir (or the datadir specified by command line argument) is for some reason inaccessible. I have created https://github.com/bitcoin/bitcoin/pull/13621 in an attempt to fix this.

**Review by ryanofsky — APPROVED** [review]
utACK 8b04bf40f7396518d638a7feacf2d8e8e9931780. The inconsistencies between doc/init.md and contrib/init/bitcoind.service and contrib/init/bitcoind.service and the other init files are confusing and this PR gets rid of them and is an improvement over the status quo.

But I think @Flowdalic makes very good points, especially about this change breaking the ability to specify datadir in the config file in https://github.com/bitcoin/bitcoin/pull/12255#issuecomment-363684576. So to follow up, I would suggest the following changes (either here or in separate PRs):

- Add release-notes-pr12255.md file mentioning that contrib/init/bitcoind.service has changed to use /var/lib/bitcoind as the datadir instead of $HOME/.bitcoin (which is typically /var/lib/bitcoin/.bitcoin or /home/bitcoin/.bitcoin). The change makes bitcoin more consistent with other services, makes the systemd config more consistent with existing upstart and openrc configs. The release notes should also mention that datadir is now managed entirely in the bitcoind.service file and that trying to override it in /etc/bitcoin/bitcoin.conf will have no effect.

- Mention somewhere in doc/init.md that it is not possible to override datadir in /etc/bitcoin/bitcoin.conf with the current systemd, openrc, and upstart init files.

- (Option for a future PR.) I think it's probably better to just document inability to override datadir in /etc/bitcoin/bitcoin.conf, rather than try to do anything about it. But if there are use-cases for wanting to use our service files but also needing the ability to change datadir from the config file, there are changes we could make to bitcoind to support removing the `-datadir=/var/lib/bitcoind` startup argument. For example we could change [GetDefaultDataDir](https://github.com/bitcoin/bitcoin/blob/d7b0258ff037ae90f60a86fc68cb55069c96e8d0/src/util/system.cpp#L726-L750) to return "$BITCOIN_HOME" if it is set, instead of "$HOME/.bitcoin", with no loss of compatibility. (GnuPG has a similar GNUPGHOME variable.) The init files could then set $BITCOIN_HOME instead of overriding -datadir.

- I think Flowdalic is right that dropping the `Group=bitcoin` line would be less fragile. This should be safe because we aren't giving the group any permissions, and [systemd.exec](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#User=) specifies "If no group is set, the default group of the user is used". But I would want the group to be specified explicitly if we were granting the group any permissions (even read only).

**Comment by dongcarl** [comment]
> I think Flowdalic is right that dropping the `Group=bitcoin` line would be less fragile. This should be safe because we aren't giving the group any permissions, and systemd.exec specifies "If no group is set, the default group of the user is used". But I would want the group to be specified explicitly if we were granting the group any permissions (even read only).

I believe that there are valid uses for the `bitcoin` group. Some examples:

- `elements` uses the bitcoin auth cookie file thru its `-mainchainrpccookiefile` flag for its RPC calls to bitcoind
- `c-lightning` allows specifying the bitcoin data directory thru its `--bitcoin-datadir` flag

Perhaps what should be done is to make the permissions for the datadir 770 or 750, so that these applications can run under the `bitcoin` group and automatically get rw/ro access to it.

**Comment by dongcarl** [comment]
> Add release-notes-pr12255.md file mentioning that contrib/init/bitcoind.service has changed to use /var/lib/bitcoind as the datadir instead of $HOME/.bitcoin (which is typically /var/lib/bitcoin/.bitcoin or /home/bitcoin/.bitcoin). The change makes bitcoin more consistent with other services, makes the systemd config more consistent with existing upstart and openrc configs. The release notes should also mention that datadir is now managed entirely in the bitcoind.service file and that trying to override it in /etc/bitcoin/bitcoin.conf will have no effect.
> 
> Mention somewhere in doc/init.md that it is not possible to override datadir in /etc/bitcoin/bitcoin.conf with the current systemd, openrc, and upstart init files.

Sounds good to me. Will do.

> (Option for a future PR.) I think it's probably better to just document inability to override datadir in /etc/bitcoin/bitcoin.conf, rather than try to do anything about it. But if there are use-cases for wanting to use our service files but also needing the ability to change datadir from the config file, there are changes we could make to bitcoind to support removing the -datadir=/var/lib/bitcoind startup argument. For example we could change GetDefaultDataDir to return "$BITCOIN_HOME" if it is set, instead of "$HOME/.bitcoin", with no loss of compatibility. (GnuPG has a similar GNUPGHOME variable.) The init files could then set $BITCOIN_HOME instead of overriding -datadir.

Just to verify I understand what you're saying:

- In the case that both `$BITCOIN_HOME` and `-datadir` are set, `-datadir` would override `$BITCOIN_HOME`
- An unset `$BITCOIN_HOME` effectively is `env BITCOIN_HOME=$HOME/.bitcoin` as it is currently

**Comment by ryanofsky** [comment]
> I believe that there are valid uses for the bitcoin group

Sure, I'm just agreeing with previous commenters that bitcoin group should either be required to exist and given permissions or it should be dropped. It's unnecessarily fragile to reference a group that doesn't need to exist and doesn't have any permissions for no reason.

> Just to verify I understand what you're saying

I'm not really saying the second thing. The [GetDefaultDataDir](https://github.com/bitcoin/bitcoin/blob/d7b0258ff037ae90f60a86fc68cb55069c96e8d0/src/util/system.cpp#L726-L750) function is more complex than that. I'm just saying we *could* add a one-line override at the top of that function:

```c++
if (char* ret = getenv("BITCOIN_HOME_DIR")) return ret;
```

that would give init systems a way to set up a default data location without causing datadir= options in /etc/bitcoin/bitcoin.conf to get silently ignored. I'm not saying we *should* do this. It's fine IMO to just document the limitations that exist right now. But if they are a problem that needs to be solved, this would be a simple, backwards compatible way to solve them.

**Comment by dongcarl** [comment]
- Cleaned up commits
- Given `bitcoin` group exec permission on directories
- Modified `doc/init.md`
- Added release note

File: contrib/init/bitcoind.service [hunk 1]
```diff
@@ -12,12 +12,11 @@ Description=Bitcoin daemon
...
 
 [Service]
-ExecStart=/usr/bin/bitcoind -daemon -conf=/etc/bitcoin/bitcoin.conf -pid=/run/bitcoind/bitcoind.pid
+ExecStart=/usr/bin/bitcoind -daemon=0 -conf=/etc/bitcoin/bitcoin.conf -pid=/run/bitcoind/bitcoind.pid
 # Creates /run/bitcoind owned by bitcoin
 RuntimeDirectory=bitcoind
 User=bitcoin
-Type=forking
-PIDFile=/run/bitcoind/bitcoind.pid
+Type=simple
```
Comment (ryanofsky) [inline contrib/init/bitcoind.service]: In commit "init: Don't daemonize when runnning under systemd" (558dddfa8a493a3fa521d7bd098ec6d87e148ced)

I don't think changing from a forking service to a simple service is actually a good idea, and I wonder if there any downsides at all to just dropping this commit, while keeping the other two commits, because:

- Simple services are more limited that forking services. You can't tell the difference between a running service and service that's starting up, and you can't do things like specifying Requires= and After= in a dependent service to avoid starting it if starting bitcoind fails.

- Using a simple service requires passing `-daemon=0` which currently has the side effect of enabling logging to stdout. If any debug options are specified this can mean a lot of information will go into system logs that isn't going there now. Also, since bitcoind prints timestamps on the console by default, the log entries will look strange in journalctl.

It might be a good idea to add `-printtoconsole=0` if we need to keep `-daemon=0`, or at least mention in release notes that new logs may now show up in journald.

**Review by ryanofsky — APPROVED** [review]
utACK 40d16e0664091e69fa921ffbf6254f4951c47ed6. I think all the issues that were brought up earlier are now addressed, and this looks like a nice improvement. I left a new comment, but it can be ignored if it's not important.

File: contrib/init/bitcoind.service (hunk 1 above)
Comment (dongcarl) [reply contrib/init/bitcoind.service]: You are right! After doing some digging I think `forking` is superior right now. Perhaps the following changes can be made in future PRs:

1. Use `sd_notify` to give `systemd` more information on the state of the process if a compile flag `--enable-systemd-notify` is supplied to `./configure`
2. Make logs available in `journald` in a sane way somehow

File: doc/release-notes/release-notes-pr12255.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+PR #12255
+=========
+
+The systemd init file (`contrib/init/bitcoind.service`) has changed to use
+`/var/lib/bitcoind` as the data directory instead of `~bitcoin/.bitcoin`. This
+change makes bitcoin more consistent with other services, and makes the systemd
```
Comment (fanquake) [inline doc/release-notes/release-notes-pr12255.md]: nit: s/has changed/has been changed
nit: s/bitcoin/Bitcoin Core

**Review by fanquake — APPROVED** [review]
utACK 3b6b72a

nits can be fixed up when the release notes are merged.

File: doc/release-notes/release-notes-pr12255.md (hunk 2 above)
Comment (dongcarl) [reply doc/release-notes/release-notes-pr12255.md]: Addressed. Thanks!

**Review by fanquake — APPROVED** [review]
re-utACK ca07fa8

**Review by ryanofsky — APPROVED** [review]
utACK ca07fa88c82928bcafd678fa4fd12e1dd4e2b1a3. Only changes since previous review are dropping the first commit (in response to https://github.com/bitcoin/bitcoin/pull/12255#discussion_r245387019) and tweaking the release notes.

File: doc/init.md [hunk 3]
```diff
@@ -65,6 +65,18 @@ reasons to make the configuration file and data directory only readable by the
...
+directories and the setting of their permissions is automatically handled by
+systemd. Directories are given a permission of 710, giving the bitcoin group
+access to files under it _if_ the files themselves give permission to the
+bitcoin group to do so (e.g. when `-sysperms` is specified). This does not allow
+for the listing of files under the directory.
+
+NOTE: It is not currently possible to override `datadir` in
+`/etc/bitcoin/bitcoin.conf` with the current systemd, OpenRC, and Upstart init
+files. The command line arguments specified in the init files take precedence
+over the configurations in `/etc/bitcoin/bitcoin.conf`.
```
Comment (luke-jr) [inline doc/init.md]: At least OpenRC's init allows for setting `BITCOIND_DATADIR` to override it.

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: doc/init.md (hunk 3 above)
Comment (ryanofsky) [reply doc/init.md]: re: https://github.com/bitcoin/bitcoin/pull/12255#discussion_r248015409

> At least OpenRC's init allows for setting BITCOIND_DATADIR to override it.

re: http://www.erisian.com.au/bitcoin-core-dev/log-2019-01-17.html#l-365

> I see... I think I'll make it easy to override for systemd as well then. Thanks!

It would be good to mention `/etc/conf.d/bitcoind` for OpenRC, but I wouldn't put a huge amount effort into implementing a configuration mechanism for the systemd service (unless your really want to). RC scripts tend to support external configuration because can they can be pretty long and complicated. Systemd service files are (in theory) supposed to stay simple and rely on more coarse override mechanisms like `systemctl edit`.

**Comment by ryanofsky** [comment]
I think this could be merged. It looks like the only outstanding suggestion is to mention the openrc config file in the documentation (https://github.com/bitcoin/bitcoin/pull/12255#discussion_r248015409), which could be done in a followup later.

**Comment by dongcarl** [comment]
@ryanofsky Perhaps I should just omit the other init systems in the release notes, as it is unrelated to this PR and I don't have enough confidence that they don't have a way of overriding. If that seems acceptable I will make the wording changes.

**Comment by ryanofsky** [comment]
The sentence:

> It is not currently possible to override `datadir` in `/etc/bitcoin/bitcoin.conf` with the current systemd, OpenRC, and Upstart init files

is correct as written. If you want to mention that OpenRC and Upstart have other config mechanisms that  would be wonderful but unnecessary. I do think it is important to mention that out of the box in all three init systems trying to set `datadir` in `/etc/bitcoin/bitcoin.conf` is not supported and will have no effect.

**Comment by dongcarl** [comment]
Pushed clearer wording

File: doc/release-notes/release-notes-pr12255.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
+PR #12255
```
Comment (laanwj) [inline doc/release-notes/release-notes-pr12255.md]: Thanks for adding a release notes item.
I'd suggest `systemd init file` as title, the PR number won't tell most users much

**Comment by dongcarl** [comment]
Fixed @laanwj's suggestion.

**Comment by laanwj** [comment]
Thanks! utACK bad1716c6d30fdf4be6d5050a04e1211f920bbd6

**Review by ryanofsky — COMMENTED** [review]
utACK bad1716c6d30fdf4be6d5050a04e1211f920bbd6
